lib/TPP/Dialect/Xsmm/XsmmUtils.cpp

```diff
@@ -287,6 +358,175 @@ FailureOr<BinaryFlags> getBinaryFlags(Type operandType, Type outputType,
+                               Operation *parentOp, bool getResults) {
+  SmallVector<Value> res;
+  builder.setInsertionPoint(parentOp);
+  for (Value operand : operands) {
+    auto memrefType = dyn_cast<MemRefType>(operand.getType());
+    if (!memrefType) {
+      res.push_back(operand);
+      continue;
+    }
+    auto [ptr, offset] = ::mlir::utils::getPtrAndOffset(builder, operand, loc);
+    res.push_back(ptr);
+    res.push_back(offset);
+  }
+  return res;
+}
+
+SmallVector<Type> extractInvokeOperandTypes(OpBuilder &builder,
+                                            ValueRange operands) {
+  SmallVector<Type> results;
+  // One extra operand for datatype
+  for (Value operand : operands) {
+    Type operandType = operand.getType();
+    if (auto memrefType = dyn_cast<MemRefType>(operandType)) {
+      // TODO: non-POD will require an LLVMTypeConverter.
+      Type basePtrType = LLVM::LLVMPointerType::get(builder.getContext());
+      results.push_back(basePtrType);
+      results.push_back(builder.getIndexType()); // offset
+    } else {
+      results.push_back(operand.getType());
+    }
+  }
+  return results;
+}
+
+int64_t getOredFlags(ArrayAttr flags) {
+  int64_t oredFlag = 0;
+  for (auto flag : flags) {
+    int64_t intAttr = dyn_cast<IntegerAttr>(flag).getInt();
+    // LIBXSMM is col-major, swap A and B flags.
+    if (auto gemmFlag = dyn_cast_or_null<xsmm::GemmFlagsAttr>(flag)) {
+      if (gemmFlag.getValue() == GemmFlags::VNNI_A)
+        intAttr = static_cast<int64_t>(GemmFlags::VNNI_B);
+      if (gemmFlag.getValue() == GemmFlags::VNNI_B)
+        intAttr = static_cast<int64_t>(GemmFlags::VNNI_A);
+    }
+    oredFlag |= intAttr;
+  }
+  return oredFlag;
+}
+
+int64_t getUnaryOredFlags(PatternRewriter &rewriter,
+                          xsmm::UnaryFlags unaryFlags) {
+  auto flags = rewriter.getArrayAttr(
+      xsmm::UnaryFlagsAttr::get(rewriter.getContext(), unaryFlags));
+  int64_t oredFlag = xsmm::utils::getOredFlags(flags);
+  return oredFlag;
+}
+
+func::CallOp buildDispatchCall(RewriterBase &rewriter, Location loc,
+                               ArrayRef<Value> dispatchOperands,
+                               ArrayRef<Type> dispatchOperandTypes,
+                               ModuleOp module, FlatSymbolRefAttr fnName) {
+  auto libFnType = rewriter.getFunctionType(
+      dispatchOperandTypes, IntegerType::get(rewriter.getContext(), 64));
+
+  if (!module.lookupSymbol(fnName.getAttr())) {
+    OpBuilder::InsertionGuard guard(rewriter);
+    // Insert before module terminator.
+    rewriter.setInsertionPoint(module.getBody(),
+                               std::prev(module.getBody()->end()));
+    func::FuncOp funcOp =
+        rewriter.create<func::FuncOp>(loc, fnName.getValue(), libFnType);
+    funcOp.setPrivate();
+  }
+
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPointAfter(dispatchOperands.back().getDefiningOp());
```

Comment:
This doesn't seem necessary. If you're building this function now, then all the producers have already been lowered into IR and the insertion point is at the right place?

Reply:
Function creation is right before, so that moves the insertion point. Thats why I think I set the insertion point explicitly, nevertheless I'll verify if I can remove this line safely.

Reply:
Removed the insertion point, as this wasnt necessary as rightly pointed out by Renato.